src/main/kotlin/com/chattriggers/ctjs/minecraft/listeners/ClientListener.kt

```diff
@@ -193,73 +195,83 @@ object ClientListener : Initializer {
             }
         }
 
-        AttackBlockCallback.EVENT.register { _, _, _, pos, direction ->
+        AttackBlockCallback.EVENT.register { player, _, hand, pos, direction ->
+            if (!player.world.isClient) return@register ActionResult.PASS
             val event = CancellableEvent()
 
             TriggerType.PLAYER_INTERACT.triggerAll(
                 PlayerInteraction.ATTACK_BLOCK,
                 World.getBlockAt(BlockPos(pos)).withFace(BlockFace.fromMC(direction)),
+                hand,
                 event,
             )
 
             if (event.isCancelled()) ActionResult.FAIL else ActionResult.PASS
         }
 
-        AttackEntityCallback.EVENT.register { _, _, _, entity, _ ->
+        AttackEntityCallback.EVENT.register { player, _, hand, entity, _ ->
+            if (!player.world.isClient) return@register ActionResult.PASS
             val event = CancellableEvent()
 
             TriggerType.PLAYER_INTERACT.triggerAll(
                 PlayerInteraction.ATTACK_ENTITY,
                 Entity.fromMC(entity),
+                hand,
                 event,
             )
 
             if (event.isCancelled()) ActionResult.FAIL else ActionResult.PASS
         }
 
-        PlayerBlockBreakEvents.BEFORE.register { _, _, pos, state, _ ->
-            val event = CancellableEvent()
-
+        CTEvents.BREAK_BLOCK.register { pos, hand ->
             TriggerType.PLAYER_INTERACT.triggerAll(
                 PlayerInteraction.BREAK_BLOCK,
-                Block(BlockType(state.block), BlockPos(pos)),
-                event,
+                World.getBlockAt(BlockPos(pos)),
+                hand,
+                CancellableEvent(),
             )
-
-            !event.isCancelled()
         }
 
-        UseBlockCallback.EVENT.register { _, _, _, hitResult ->
+        UseBlockCallback.EVENT.register { player, _, hand, hitResult ->
+            // TODO: double firing
```

Comment:
To clarify, it double fires event with the client check?

Reply:
Oops. Meant to delete that comment. The double firing was from triggering on both the main hand and offhand.